src/peer.ts

```diff
@@ -1,15 +1,194 @@
-import { IpRegistry } from "./ipRegistry";
+import { logger } from "@blockr/blockr-logger";
+import { Guid } from "guid-typescript";
+
+import { MessageType } from "./enums";
+import { IMessageListener } from "./interfaces/iMessageListener";
+import { IPeer } from "./interfaces/iPeer";
+import { Message } from "./models/message";
 import { Receiver } from "./receiver";
+import { RoutingTable } from "./routingTable";
 import { Sender } from "./sender";
+import { DateManipulator } from "./util/dateManipulator";
+
+const MESSAGE_EXPIRATION_TIMER: number = 1;
+const MESSAGE_HISTORY_CLEANUP_TIMER: number = 60000; // One minute
+
+/**
+ * Handles the peer network.
+ */
+export class Peer implements IMessageListener, IPeer {
+    private readonly routingTable: RoutingTable;
+    private readonly receiveHandlers: Map<string, (message: Message, senderIp: string) => void>;
+    private readonly sender: Sender;
+    private readonly receiver: Receiver;
+    private GUID: string;
+
+    constructor(port: string, initialPeers?: string[]) {
+        this.routingTable = new RoutingTable();
+        this.receiveHandlers = new Map();
+        this.createReceiverHandlers();
+
+        this.sender = new Sender(port);
+        this.receiver = new Receiver(this, port);
+
+        // Create timer that removes peers that did not reply
+        setInterval(() => {
+            const minDate = DateManipulator.minusMinutes(new Date(), MESSAGE_EXPIRATION_TIMER);
+            this.sender.getSentMessagesSendersSince(minDate).forEach((value: string) => {
+                this.routingTable.removePeer(value);
+                logger.info(`Peer removed from routing table: ${value}`);
+            });
+        }, MESSAGE_HISTORY_CLEANUP_TIMER);
+
+        // If initialPeers is undefined, this instance is the first peer
+        if (initialPeers === undefined) {
+            this.GUID = Guid.create().toString();
+            return;
+        }
+        this.GUID = Guid.createEmpty().toString();
+        this.checkInitialPeers(initialPeers);
+    }
+
+    /**
+     * Register custom receiver handlers.
+     *
+     * @param messageType - The messageType that the receiver handles
+     * @param implementation - The implementation of the receiver handler
+     */
+    public registerReceiveHandlerImpl(messageType: string, implementation: (message: Message, senderIp: string) => void): void {
+        this.receiveHandlers.set(messageType, implementation);
+    }
+
+    /**
+     * Send a message to the given destination.
+     *
+     * @param messageType - The message type
+     * @param destination
+     * @param [body] - The message body
+     */
+    public sendMessage(messageType: string, destination: string, body?: string): void {
+        const destinationIp = this.routingTable.peers.get(destination);
+        if (destinationIp === undefined) {
+            throw new Error(`Unknown destination. Could not find an IP for: ${destination}`);
```

Comment:
Maybe create a custom UnknownDestinationError instead of using the standard Error class.